// react-pages/modules/components/Loading.js
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import classNames from 'classnames';
import LoadingIndicator from './LoadingIndicator';
import FadeInOut from './FadeInOut';
export function Loading(_ref) {
  var initial = _ref.initial,
      pending = _ref.pending,
      immediate = _ref.immediate,
      Indicator = _ref.indicator,
      fadeOutDuration = _ref.fadeOutDuration;
  return React.createElement("div", {
    className: classNames('rrui__fixed-full-width', 'react-pages__loading', {
      'react-pages__loading--initial': initial,
      'react-pages__loading--shown': pending,
      'react-pages__loading--immediate': immediate
    })
  }, React.createElement(FadeInOut, {
    show: pending,
    fadeOutDuration: fadeOutDuration
  }, React.createElement(Indicator, {
    className: "react-pages__loading-spinner"
  })));
}
Loading.propTypes = {
  initial: PropTypes.bool,
  pending: PropTypes.bool,
  immediate: PropTypes.bool,
  indicator: PropTypes.func.isRequired,
  fadeOutDuration: PropTypes.number.isRequired
};
Loading.defaultProps = {
  indicator: LoadingIndicator,
  fadeOutDuration: 160
};
export default connect(function (_ref2) {
  var preload = _ref2.preload;
  return {
    initial: preload.initial,
    pending: preload.pending,
    immediate: preload.immediate
  };
})(Loading);
