// react-pages/commonjs/static-site-generator/upload.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = upload_website;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _progress = _interopRequireDefault(require("progress"));

// const green = '\u001b[42m \u001b[0m'
function upload_website(_x, _x2) {
  return _upload_website.apply(this, arguments);
}

function _upload_website() {
  _upload_website = (0, _asyncToGenerator2["default"])(
  /*#__PURE__*/
  _regenerator["default"].mark(function _callee(filesPath, upload) {
    var progressBar;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return upload(filesPath, {
              started: function started(total) {
                progressBar = new _progress["default"]('  Uploading [:bar] :percent :etas', {
                  // complete : green,
                  complete: '=',
                  incomplete: ' ',
                  width: 50,
                  total: 100
                });
              },
              progress: function progress(value) {
                progressBar.update(value);
              }
            });

          case 2:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _upload_website.apply(this, arguments);
}
